File: common/src/main/java/me/codexadrian/tempad/common/data/LocationData.java
package me.codexadrian.tempad.common.data;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import com.teamresourceful.bytecodecs.base.ByteCodec;
import com.teamresourceful.bytecodecs.base.object.ObjectByteCodec;
import com.teamresourceful.resourcefullib.common.codecs.CodecExtras;
import com.teamresourceful.resourcefullib.common.utils.CommonUtils;
import me.codexadrian.tempad.common.Tempad;
import me.codexadrian.tempad.common.utils.CodecUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.UUIDUtil;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtOps;
import net.minecraft.nbt.NbtUtils;
import net.minecraft.nbt.Tag;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.ExtraCodecs;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.UUID;

public record LocationData(String name, ResourceKey<Level> levelKey, BlockPos blockPos, float angle, UUID id, boolean isDeletable, boolean isDownloadable) implements Comparable<LocationData> {
    public static ByteCodec<LocationData> BYTE_CODEC = ObjectByteCodec.create(
        ByteCodec.STRING.fieldOf(LocationData::name),
        CodecUtils.DIMENSION.fieldOf(LocationData::levelKey),
        CodecUtils.BLOCK_POS.fieldOf(LocationData::blockPos),
        ByteCodec.FLOAT.fieldOf(LocationData::angle),
        ByteCodec.UUID.fieldOf(LocationData::id),
        ByteCodec.BOOLEAN.fieldOf(LocationData::isDeletable),
        ByteCodec.BOOLEAN.fieldOf(LocationData::isDownloadable),
        LocationData::new
    );

    public static Codec<LocationData> CODEC = RecordCodecBuilder.create(
        instance -> instance.group(
            Codec.STRING.fieldOf("name").forGetter(LocationData::name),
            Level.RESOURCE_KEY_CODEC.fieldOf("levelKey").forGetter(LocationData::levelKey),
            BlockPos.CODEC.fieldOf("blockpos").forGetter(LocationData::blockPos),
            Codec.FLOAT.optionalFieldOf("angle", 0f).forGetter(LocationData::angle),
            UUIDUtil.CODEC.fieldOf("uuid").forGetter(LocationData::id),
            Codec.BOOL.optionalFieldOf("isDeletable", true).forGetter(LocationData::isDeletable),
            Codec.BOOL.optionalFieldOf("isDownloadable", true).forGetter(LocationData::isDownloadable)
        ).apply(instance, LocationData::new)
    );

    public LocationData(String name, @Nullable ResourceKey<Level> levelKey, BlockPos pos, float angle) {
        this(name, levelKey, pos, angle, UUID.randomUUID());
    }

    public LocationData(String name, @Nullable ResourceKey<Level> levelKey, BlockPos pos, float angle, UUID uuid) {
        this(name, levelKey, pos, angle, uuid, true, true);
    }

    public LocationData(String name, @Nullable ResourceKey<Level> levelKey, BlockPos pos) {
        this(name, levelKey, pos, 0, UUID.randomUUID(), true, true);
    }

    public LocationData(String name, @Nullable ResourceKey<Level> levelKey, BlockPos pos, UUID uuid) {
        this(name, levelKey, pos, 0, uuid, true, true);
    }

    public CompoundTag toTag() {
        CompoundTag tag = new CompoundTag();
        tag.putUUID("uuid", id);
        tag.putString("name", name);
        tag.put("blockpos", NbtUtils.writeBlockPos(blockPos));
        Level.RESOURCE_KEY_CODEC.encodeStart(NbtOps.INSTANCE, levelKey).result().ifPresent(levelKey -> tag.put("levelKey", levelKey));

        return tag;
    }

    public static LocationData fromTag(CompoundTag tag) {
        UUID uuid = tag.getUUID("uuid");
        String name = tag.getString("name");
        BlockPos pos = NbtUtils.readBlockPos(tag.getCompound("blockpos"));
        ResourceKey<Level> key = Level.RESOURCE_KEY_CODEC.parse(NbtOps.INSTANCE, tag.get("levelKey")).result().orElse(null);

        return new LocationData(name, key, pos, uuid);
    }

    @Override
    public int compareTo(@NotNull LocationData o) {
        return name().compareTo(o.name());
    }
}

File: src/main/kotlin/earth/terrarium/tempad/api/locations/LocationData.kt
package earth.terrarium.tempad.api.locations

import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import com.teamresourceful.bytecodecs.base.ByteCodec
import com.teamresourceful.bytecodecs.base.`object`.ObjectByteCodec
import com.teamresourceful.resourcefullib.common.bytecodecs.ExtraByteCodecs
import com.teamresourceful.resourcefullib.common.color.Color
import earth.terrarium.tempad.common.utils.COLOR_BYTE_CODEC
import earth.terrarium.tempad.common.utils.VEC3_BYTE_CODEC
import earth.terrarium.tempad.common.utils.byteCodec
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.ComponentSerialization
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.Mth
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3
import thedarkcolour.kotlinforforge.neoforge.forge.vectorutil.v3d.plus

data class LocationType<T: NamedGlobalLocation>(val id: ResourceLocation, val byteCodec: ByteCodec<T>, val codec: Codec<T>)

interface NamedGlobalLocation {
    val name: String
    val pos: Vec3
    val dimension: ResourceKey<Level>
    val angle: Float
    val color: Color

    val display: List<Component>
    val type: LocationType<*>
}

object LocationTypeRegistry {

}

data class LocationData(val name: String, val pos: Vec3, val dimension: ResourceKey<Level>, val angle: Float, val color: Color) {
    companion object {
        val MAP_CODEC: MapCodec<LocationData> = RecordCodecBuilder.mapCodec {
            it.group(
                Codec.STRING.fieldOf("name").forGetter { it.name },
                Vec3.CODEC.fieldOf("pos").forGetter { it.pos },
                Level.RESOURCE_KEY_CODEC.fieldOf("dimension").forGetter { it.dimension },
                Codec.FLOAT.fieldOf("angle").forGetter { it.angle },
                Color.CODEC.fieldOf("color").forGetter(LocationData::color),
            ).apply(it, ::LocationData)
        }

        val CODEC: Codec<LocationData> = MAP_CODEC.codec()

        val BYTE_CODEC: ByteCodec<LocationData> = ObjectByteCodec.create(
            ByteCodec.STRING.fieldOf { it.name },
            VEC3_BYTE_CODEC.fieldOf { it.pos },
            ExtraByteCodecs.DIMENSION.fieldOf { it.dimension },
            ByteCodec.FLOAT.fieldOf { it.angle },
            COLOR_BYTE_CODEC.fieldOf { it.color },
            ::LocationData
        )

        fun offsetLocation(pos: Vec3, angle: Float, distance: Int = 1): Vec3 {
            val angleInRadians = (angle + 90) * Mth.DEG_TO_RAD
            return pos + Vec3(Mth.cos(angleInRadians).toDouble() * distance, 0.0, Mth.sin(angleInRadians).toDouble() * distance)
        }
    }

    val offsetLocation = offsetLocation(pos, angle + 180)

    val x: Int = pos.x.toInt()
    val y: Int = pos.y.toInt()
    val z: Int = pos.z.toInt()
    val dimensionText = Component.translatable(dimension.location().toLanguageKey("dimension"))

    val dimComponent: Component = Component.translatable(dimension.location().toLanguageKey("dimension"))
}